sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/ApproximateDistinct.java

```diff
@@ -0,0 +1,533 @@
+   * {@link ApproximateDistinctFn} as a {@link org.apache.beam.sdk.state.CombiningState} :
+   *
+   * <pre>{@code
+   * class StatefulCardinality<V> extends DoFn<V>, OutputT>> {
+   *   {@literal @}StateId("hyperloglog")
+   *   private final StateSpec<CombiningState<V, HyperLogLogPlus, HyperLogLogPlus>> indexSpec;
+   *
+   *   public StatefulCardinality(ApproximateDistinctFn<V> fn) {
+   *     indexSpec = StateSpecs.combining(fn);
+   *   }
+   *
+   *   {@literal @}ProcessElement
+   *   public void processElement(
+   *      ProcessContext context,
+   *      {@literal @}StateId("hllSketch")
+   *      CombiningState<V, HyperLogLogPlus, HyperLogLogPlus> hllSketch) {
+   *     long current = MoreObjects.firstNonNull(hllSketch.getAccum().cardinality(), 0L);
+   *     hllSketch.add(context.element());
+   *     context.output(...);
+   *     }
+   * }
+   * }</pre>
+   *
+   * <p>Then the {@link org.apache.beam.sdk.transforms.DoFn} can be called like this :
+   * <pre>{@code
+   * PCollection<Object> input = ...;
+   * ApproximateDistinctFn myFn = ApproximateDistinctFn.create(new ObjectCoder());
+   * PCollection<OutputT> = input.apply(ParDo.of(new StatefulCardinality(myFn)));
+   * }</pre>
+   *
+   * @param <InputT>      the type of the elements in the input {@link PCollection}
+   */
+  public static class ApproximateDistinctFn<InputT>
+      extends CombineFn<InputT, HyperLogLogPlus, HyperLogLogPlus> {
+
+    private final int p;
+
+    private final int sp;
+
+    private final Coder<InputT> inputCoder;
+
+    private ApproximateDistinctFn(int p, int sp, Coder<InputT> coder) {
+      this.p = p;
+      this.sp = sp;
+      inputCoder = coder;
+    }
+
+    /**
+     * Returns an {@link ApproximateDistinctFn} combiner with the given input coder.
+     *
+     * @param <InputT>    the type of the elements in the input {@link PCollection}
+     * @param coder       the coder that encodes the elements' type
+     */
+    public static <InputT> ApproximateDistinctFn<InputT> create(Coder<InputT> coder) {
+      try {
+        coder.verifyDeterministic();
+      } catch (Coder.NonDeterministicException e) {
+        throw new IllegalArgumentException("Coder is not deterministic ! " + e.getMessage(), e);
+      }
+      return new ApproximateDistinctFn<>(12, 0, coder);
+    }
+
+    /**
+     * Returns a new {@link ApproximateDistinctFn} combiner with a new precision p but
+     * the other parameters remain unchanged.
+     *
+     * @param p           the precision value for the normal representation
+     */
+    public ApproximateDistinctFn<InputT> withPrecision(int p) {
+      if (p < 4) {
+        throw new IllegalArgumentException(String.format("Expected : p >= 4. Actual : p = %s", p));
+      }
+      return new ApproximateDistinctFn<>(p, this.sp, this.inputCoder);
+    }
+
+    /**
+     * Returns a new {@link ApproximateDistinctFn} combiner with a sparse representation
+     * of precision sp but the other parameters remain unchanged.
```

Comment:
Generally "other parameters remain unchanged" is not worth mentioning: all Beam transform builders are this way. I'd say just `/** Uses sparse representation with the given precision. See ... for an explanation */`